[runtime env] Local uri caching for working_dir, py_modules and conda

architkulkarni opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
Previously, local files corresponding to runtime env URIs were eagerly garbage collected as soon as there were no more references to them.  In this PR, we store this data in a cache instead, so when the reference count for a URI drops to zero, instead of deleting it we simple mark it as unused in the cache.  When the cache exceeds its size limit (default 10 GB) it will delete unused URIs until the cache is back under the size limit or there are no more unused URIs.

Design doc: https://docs.google.com/document/d/1x1JAHg7c0ewcOYwhhclbuW0B0UC7l92WFkF4Su0T-dk/edit

- Adds unit tests for caching and integration tests for working_dir caching


## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

architkulkarni commented:
Failures:
- Windows job broken on master
- tune: test_commands flaky on master
- test_component_failures_2 broken on master
- test_output broken on master
- privileged test flaky on master
- Dashboard+Serve tests broken on master

rkooo567 commented:
Is this reviewable? Can you resolve the merge conflict before that?

SongGuyang commented:
@architkulkarni Can you rebase master? Do you have any plan for this PR?

architkulkarni commented:
> @architkulkarni Can you rebase master? Do you have any plan for this PR?

Yes, I will rebase master and try to get this merged as soon as possible.

edoakes commented:
@architkulkarni FYI more merge conflicts :(

edoakes commented on dashboard/modules/runtime_env/runtime_env_agent.py:
```diff
@@ -42,6 +53,79 @@ class CreatedEnvResult:
     result: str
 
 
+class URICache:
```
Can we move this to a separate file in `_private/runtime_env` please?

edoakes replied:
Please also add a docstring for this class with what it does + semantics

edoakes commented on dashboard/modules/runtime_env/runtime_env_agent.py:
```diff
@@ -102,17 +196,97 @@ def run_setup_with_logger():
+                # Set up conda
+                conda_resources_dir = os.path.join(self._runtime_env_dir,
+                                                   "conda")
+                try_to_create_directory(conda_resources_dir)
+                conda_file_lock = os.path.join(self._runtime_env_dir,
+                                               "ray-conda-install.lock")
+                # TODO(architkulkarni): add comment
+                with FileLock(conda_file_lock):
```
hmm this part smells really funny to me -- why is it necessary to do it here and not in the plugin class?

architkulkarni replied:
I think we want (1) checking the URI cache, (2) installing the conda env and (3) updating the URI cache to all be under a single lock.  Otherwise we get a race condition when two workers both need the same conda env and see an empty URI cache.  But I'm a little shaky on the best practices for locking, maybe you know a better way?

architkulkarni replied:
I see, I guess the alternative is to do this: 
Inside the plugin class, change `create_conda_env` to `create_conda_env_if_needed`, and only put a lock around the call `create_conda_env_if_needed`.  

I was originally concerned about how to decide the "if_needed" part, because if you just check the existence of the conda env path, you can't tell whether the env is fully installed or if it's still in progress.  But if the call to `create_conda_env_if_needed` is wrapped in a lock, then only one process can access it at a time, so we won't have this problem.

edoakes commented on python/ray/_private/runtime_env/packaging.py:
```diff
@@ -166,6 +166,8 @@ def parse_uri(pkg_uri: str) -> Tuple[Protocol, str]:
             -> ("gs",
             "gs_public-runtime-env-test_test_module.zip")
     """
+    print("PARSE URI: ", pkg_uri)
```
remove

edoakes commented on python/ray/_private/runtime_env/pip.py:
```diff
@@ -77,21 +85,27 @@ def delete_uri(self,
+                     f"{runtime_env.serialize()}")
+        protocol, hash = parse_uri(uri)
+        target_dir = self._get_path_from_hash(hash)
+        # TODO(architkulkarni): Decide what to do about this
+        # if not runtime_env.has_pip():
+        #     return
```
what's the context here?

edoakes commented on python/ray/_private/runtime_env/pip.py:
```diff
@@ -77,21 +85,27 @@ def delete_uri(self,
-        logger.debug(f"Setting up pip for runtime_env: {runtime_env}")
         pip_packages: List[str] = runtime_env.pip_packages()
-        target_dir = self._get_path_from_hash(_get_pip_hash(pip_packages))
+        # target_dir = self._get_path_from_hash(_get_pip_hash(pip_packages))
```
?

architkulkarni commented:
Thanks for the feedback @edoakes, it's still a work in progress but I'll let you know when it's ready for review.

architkulkarni commented:
Tests passed in CI, I'll do some cleanup and address the preliminary comments and then I'll ask for reviews

architkulkarni commented:
@edoakes Tests are passing, ready for review

architkulkarni reviewed (commented):
The diffs of the test files in this PR are messed up and hard to read.  A summary of the changes: 

- Add `runtime_env_disable_URI_cache` fixture in existing GC tests
- Add new tests `test_default_large_cache` and `test_hit_cache_size_limit`

edoakes commented:
Let's have @rkooo567 review as well for a second pair of eyes

edoakes commented on dashboard/modules/runtime_env/runtime_env_agent.py:
```diff
@@ -102,17 +123,50 @@ def run_setup_with_logger():
+                            uri_cache.mark_used(uri, logger=per_job_logger)
+                    manager.modify_context(uri, runtime_env, context)
+
+                # Set up py_modules. For now, py_modules uses multiple URIs so
+                # the logic is slightly different from working_dir, conda, and
+                # pip above.
+                py_modules_uris = self._py_modules_manager.get_uris(
```
what is the plan for unifying this>?

architkulkarni replied:
In a followup we should use a single URI in the agent.  The py_modules plugin itself will manage the "sub-uris" representing individual modules, but the agent won't know about this.  (This is the same strategy we had in mind for the Anyscale proprietary working_dir, Proposal 2 here: https://docs.google.com/document/d/18X8rx6Q-vYj6FyYWQEjqqnH0KRx97-OrTDuw7NwV5PM/edit#heading=h.pocjsgysu6xf )

architkulkarni replied:
In a followup we should use a single URI in the agent. The py_modules plugin itself will manage the "sub-uris" representing individual modules, but the agent won't know about this.  (This is similar to what we had discussed earlier for a more complicated working_dir plugin using multiple URIs)

edoakes replied:
this makes sense

edoakes commented on dashboard/modules/runtime_env/runtime_env_agent.py:
```diff
@@ -102,17 +123,50 @@ def run_setup_with_logger():
+                                              self._pip_uri_cache)]:
+                    uri = manager.get_uri(runtime_env)
+                    if uri is not None:
+                        if uri not in uri_cache:
+                            size_bytes = manager.create(
+                                uri,
+                                runtime_env,
+                                context,
+                                logger=per_job_logger)
+                            uri_cache.add(
+                                uri, size_bytes, logger=per_job_logger)
+                        else:
+                            uri_cache.mark_used(uri, logger=per_job_logger)
```
can we add log statements for cache hits & cache misses here? can be DEBUG level if we think it'll be spammy, but it'll help debugging a lot

edoakes commented on doc/source/handling-dependencies.rst:
```diff
@@ -335,8 +335,13 @@ The ``runtime_env`` is a Python dictionary including one or more of the followin
+
+When the cache size limit is exceeded, resources not currently used by any actor, task or job will be deleted.
+
+To disable all deletion behavior (for example, for debugging purposes) you may set the environment variable ``RAY_runtime_env_skip_local_gc`` to ``1`` on each node in your cluster before starting Ray.
```
```suggestion
To disable all deletion behavior (for example, for debugging purposes) you may set the environment variable ``RAY_runtime_env_skip_local_gc`` to ``1`` on each node in your cluster before starting Ray.
```
```suggestion
To disable all deletion behavior (for example, for debugging purposes) you may set the environment variable ``RAY_RUNTIME_ENV_SKIP_LOCAL_GC`` to ``1`` on each node in your cluster before starting Ray.
```

edoakes commented on doc/source/handling-dependencies.rst:
```diff
@@ -335,8 +335,13 @@ The ``runtime_env`` is a Python dictionary including one or more of the followin
 
-**Garbage Collection**.  Runtime environment resources on each node (such as conda environments, pip packages, or downloaded ``working_dir`` or ``py_modules`` folders) will be removed when they are no longer
-referenced by any actor, task or job.  To disable this (for example, for debugging purposes) set the environment variable ``RAY_runtime_env_skip_local_gc`` to ``1`` on each node in your cluster before starting Ray (e.g. with ``ray start``).
+Caching and Garbage Collection
```
nice, this section is very clear.

edoakes commented on python/ray/_private/runtime_env/working_dir.py:
```diff
@@ -76,33 +76,51 @@ class WorkingDirManager:
     def __init__(self, resources_dir: str):
         self._resources_dir = os.path.join(resources_dir, "working_dir_files")
         try_to_create_directory(self._resources_dir)
+        # TODO(architkulkarni): This dict is not necessary, we should use the
+        # same URI parsing function used elsewhere to convert URIs to paths.
```
Is it needed for the conda and pip cases?

edoakes commented on python/ray/_private/utils.py:
```diff
@@ -1156,3 +1156,16 @@ def check_dashboard_dependencies_installed() -> bool:
         return False
+
+
+def get_directory_size(path="."):
```
add type hints please

edoakes reviewed (commented):
This looks good to me. The two missing pieces / follow-ups I think we need now are:

- Recovery logic so we can handle the case when the agent crashes and is restarted. Right now, I'm not even sure what would happen. We would probably just error due to unexpected files? A simple fix here would be to delete everything in the runtime_resources dir on startup (purge the caches). That would at least give us correctness, and salvaging the cache in that condition is an edge case.
- Ideally, unifying all of the plugins into the same class interface to avoid all of the duplicated code. This is pretty low priority.

architkulkarni commented:
> This looks good to me. The two missing pieces / follow-ups I think we need now are:
> 
> * Recovery logic so we can handle the case when the agent crashes and is restarted. Right now, I'm not even sure what would happen. We would probably just error due to unexpected files? A simple fix here would be to delete everything in the runtime_resources dir on startup (purge the caches). That would at least give us correctness, and salvaging the cache in that condition is an edge case.
> * Ideally, unifying all of the plugins into the same class interface to avoid all of the duplicated code. This is pretty low priority.

Thanks for the review! I can handle these these in followup PRs.  I'll need to think more about how to handle agent restarts but I think your idea makes sense.  In the current state, I think our conda manager handles existing envs gracefully, but pip might error if the package already exists in the target dir.

@rkooo567 Could you please review this PR when you get a chance?  Sorry for the delay.

rkooo567 commented:
Looking at it now!

rkooo567 commented on python/ray/_private/runtime_env/conda_utils.py:
```diff
@@ -92,18 +92,24 @@ def create_conda_env(conda_yaml_file: str,
-            raise RuntimeError(
-                f"Failed to install conda environment {prefix}:\n{output}")
+    logger.info(f"Creating conda environment {prefix}")
+    exit_code, output = exec_cmd_stream_to_logger(create_cmd, logger)
```
Sorry for the unrelated message; But does it block the main thread of agents? This seems to be dangerous if that's the case

architkulkarni replied:
Good question, it's running in a separate thread: https://github.com/ray-project/ray/blob/master/dashboard/modules/runtime_env/runtime_env_agent.py#L148

rkooo567 replied:
good to know! And glad it is already running in a separate thread :)

rkooo567 replied:
```
Changed in version 3.5: If max_workers is None or not given, it will default to the number of processors on the machine, multiplied by 5, assuming that ThreadPoolExecutor is often used to
```

So, we start num_cpus * 5 default executors, which I think is pretty good for runtime env (IO heavy)

rkooo567 commented on python/ray/_private/runtime_env/conda.py:
```diff
@@ -254,60 +268,55 @@ def delete_uri(self,
-              runtime_env: RuntimeEnv,
-              context: RuntimeEnvContext,
-              logger: Optional[logging.Logger] = default_logger):
+    def create(self,
```
Maybe we should create an abstract class for all Manager?

rkooo567 commented on python/ray/_private/runtime_env/packaging.py:
```diff
@@ -416,6 +416,13 @@ def upload_package_if_needed(
 
+def get_destination_dir_from_uri(uri: str, base_directory: str) -> Path:
+    """Return the local directory to which this URI should be unpacked."""
+    pkg_file = Path(_get_local_path(base_directory, uri))
```
Is this related to this PR?

architkulkarni replied:
It's needed for `modify_context` for `working_dir`.  We need to be able to get the name of the working_dir based on the URI so that we can append `cd working_dir_name` to the worker startup command.

rkooo567 commented on python/ray/_private/runtime_env/pip.py:
```diff
@@ -64,54 +70,75 @@ def _get_path_from_hash(self, hash: str) -> str:
         return os.path.join(self._resources_dir, hash)
 
+    def get_uri(self, runtime_env: RuntimeEnv) -> Optional[str]:
+        """Return `"conda://<hash_dependencies>"` or None if no GC required."""
```
Doctstring seems wrong?

rkooo567 commented on python/ray/_private/runtime_env/conda.py:
```diff
@@ -243,6 +250,13 @@ def _get_path_from_hash(self, hash: str) -> str:
         return os.path.join(self._resources_dir, hash)
 
+    def get_uri(self, runtime_env: RuntimeEnv) -> Optional[str]:
+        """Return `"conda://<hash_dependencies>"` or None if no GC required."""
```
The docstring seems wrong? Isn't it returning None if conda is not required by the runtime env?

rkooo567 commented on python/ray/_private/runtime_env/uri_cache.py:
```diff
@@ -0,0 +1,105 @@
+        while (self._unused_uris
+               and self.get_total_size_bytes() > self.max_total_size_bytes):
+            # TODO(architkulkarni): Evict least recently used URI instead
+            arbitrary_unused_uri = next(iter(self._unused_uris))
```
What's the order policy? Is it by the random order?

architkulkarni replied:
Currently it's in a random order.  But in a followup we plan to change it to LRU

rkooo567 replied:
sounds good to me!

rkooo567 commented on python/ray/_private/runtime_env/uri_cache.py:
```diff
@@ -0,0 +1,105 @@
+            self._delete_fn(arbitrary_unused_uri, logger)
+
+    def _check_valid(self):
+        assert self._used_uris.keys() & self._unused_uris.keys() == set()
```
I think we don't need it? Maybe setting it only when debug mode is on or sth?

rkooo567 commented on python/ray/_private/runtime_env/uri_cache.py:
```diff
@@ -0,0 +1,105 @@
+        """Add a URI to the cache and mark it as in use."""
+        if uri in self._unused_uris:
+            if size_bytes != self._unused_uris[uri]:
+                logger.warning(f"Added URI {uri} with size {size_bytes}, which"
```
Is this warning really useful? Is it a bug that size bytes are changed?

architkulkarni replied:
URIs are content hashes, so we don't expect the size to change.  Let me change this to a debug log and add a comment explaining it

rkooo567 replied:
Can we just assert in this case?

rkooo567 replied:
Also, I feel like if the content size is not expected to be changed. Do we even need to track of size of used/unused uris?

Why don't we make both used/unused as a set and keep track of the total uri sizes in a separate integer attribute?

rkooo567 replied:
Feel like this will make code a lot simpler

architkulkarni replied:
If we only keep track of one integer `total_uri_size`, then how can we update `total_uri_size` when changing a particular URI from used to unused, or vice versa?  Don't we need to store a map from URIs to their sizes?

architkulkarni replied:
> Why don't we just assert in this case?

That sounds fine to me.

rkooo567 replied:
> If we only keep track of one integer total_uri_size, then how can we update total_uri_size when changing a particular URI from used to unused, or vice versa? Don't we need to store a map from URIs to their sizes?

It seems like the current API only cares the sum of used + unused. And it is guaranteed one uri is only in one of them; doesn't this mean we don't need this -> `update total_uri_size when changing a particular URI from used to unused`? I might be missing something though

rkooo567 replied:
yes that's right! So we can remove this per-uri size tracking logic since it doesn't seem to be unnecessary (even after we introduce LRU)

rkooo567 commented on python/ray/tests/test_runtime_env.py:
```diff
@@ -279,6 +281,66 @@ def f():
+        assert "big" not in cache
+        assert cache.get_total_size_bytes() == 3
+
+    def test_many_URIs(self):
```
I think we need a test to verify delete_fn is called when the size exceeds

architkulkarni replied:
Great idea, added!

rkooo567 commented on doc/source/ray-core/handling-dependencies.rst:
````diff
@@ -335,8 +335,13 @@ The ``runtime_env`` is a Python dictionary including one or more of the followin
-referenced by any actor, task or job.  To disable this (for example, for debugging purposes) set the environment variable ``RAY_runtime_env_skip_local_gc`` to ``1`` on each node in your cluster before starting Ray (e.g. with ``ray start``).
+Caching and Garbage Collection
+""""""""""""""""""""""""""""""
+Runtime environment resources on each node (such as conda environments, pip packages, or downloaded ``working_dir`` or ``py_modules`` directories) will be cached on the cluster to enable quick reuse across different tasks and actors within a job.  Each field of ``runtime_env``` has its own cache size which defaults to 10 GB.  To change this default, you may set the environment variable ``RAY_RUNT ...
````
> Each field of ``runtime_env``` has its own cache size which defaults to 10 GB.

I am a bit confused what this means. Does it mean if we have 10 runtime env, we will have 100GB caches?

architkulkarni replied:
This is valuable feedback, thanks!  I'll edit to make it more clear.  Basically, even if you have 10 runtime_envs, 10GB will be dedicated to the working_dir cache, 10Gb will be dedicated to the `pip` cache, and so on.

architkulkarni replied:
Updated the doc; please let me know if it's still unclear

rkooo567 replied:
Hmm why do we have it per runtime env type? This seems a bit odd to me. I think one of the requirements of this feature is to control disk usage, and it seems to be unintuitive.

architkulkarni replied:
I see. @rkooo567 should we change the design to just have a single unified cache for all fields?  I don't have a concrete use case in mind for why users would need different cache sizes for, say, `working_dir` and `pip`.  It's more flexible, but I think you're right that it comes at the expense of usability.

rkooo567 replied:
Yeah to me, it seems weird we have per-runtime-env cache (It is also conceptually more confusing). But if this was the design choice that was made during design review, I can live with that. cc @edoakes should we discuss this?

rkooo567 reviewed (commented):
I left some initial comments; 

High level comment is; 

- The manager class APIs seem a bit brittle (especially with cache). Can we clean this up with ABCClass? I also think we should hide cache behind the manager's create/delete APIs.
- Cache means the agent becomes stateful, which means the behavior after agent restarts is undefined. Is there a design for how to handle this? Also, if we don't handle this right now, can we make this feature off by default? We shouldn't have non-fully working feature just for the sake of performance optimization, especially when the feature is already GA

edoakes commented:
@rkooo567 for your question re: handling agent failures. For now, we can simply delete all files in the directory (clear the cache) when the agent starts up

rkooo567 commented:
> @rkooo567 for your question re: handling agent failures. For now, we can simply delete all files in the directory (clear the cache) when the agent starts up

Currently in worker pool, does it not know about if uri exists or not? It seems to only work if worker pool doesn't know the existence of uri, but only requests to create/delete them. 

Also, why don't we feature flag this for now and enable the feature after we implement this?

architkulkarni commented on doc/source/ray-core/handling-dependencies.rst:
````diff
@@ -335,8 +335,13 @@ The ``runtime_env`` is a Python dictionary including one or more of the followin
-referenced by any actor, task or job.  To disable this (for example, for debugging purposes) set the environment variable ``RAY_runtime_env_skip_local_gc`` to ``1`` on each node in your cluster before starting Ray (e.g. with ``ray start``).
+Caching and Garbage Collection
+""""""""""""""""""""""""""""""
+Runtime environment resources on each node (such as conda environments, pip packages, or downloaded ``working_dir`` or ``py_modules`` directories) will be cached on the cluster to enable quick reuse across different tasks and actors within a job.  Each field of ``runtime_env``` has its own cache size which defaults to 10 GB.  To change this default, you may set the environment variable ``RAY_RUNT ...
````
I'm not sure what qualifies as an official API.  We weren't planning to make a proposal, is it required?  We have a design document [here](https://docs.google.com/document/d/1x1JAHg7c0ewcOYwhhclbuW0B0UC7l92WFkF4Su0T-dk/edit) which was approved in the Architecture review

rkooo567 replied:
It is not required. I am fine with env var!

architkulkarni commented:
Yup, the worker pool doesn't know the existence of URIs, it just asks the agent to create and delete them like you said.  The only place where we would keep track of the existence of URIs is the agent.

I don't think agent restarts are handled well prior to this PR, so this PR would be an improvement not a regression.  But feature flagging it for now sounds fine to me.  I agree with the high level comments about using an abstract base class and moving the cache into this class.  

If I add a feature flag, would this PR be mergeable @rkooo567?  I could handle the refactoring and the agent restart handling in future PRs.  If not, I'll do them in this PR.

rkooo567 commented:
> If I add a feature flag, would this PR be mergeable @rkooo567? I could handle the refactoring and the agent restart handling in future PRs. If not, I'll do them in this PR.

I think we can merge it if we feature flag it off. That seems better for the long term too. 

> Yup, the worker pool doesn't know the existence of URIs, it just asks the agent to create and delete them like you said. The only place where we would keep track of the existence of URIs is the agent.

Isn't the previous behavior at least correct? My guess is we will just unnecessarily re-create runtime env when agent restarts, but there's no correctness issue whereas the new protocol will have the correctness issue? (e.g., URIs will never be deleted).

architkulkarni commented:
> > Yup, the worker pool doesn't know the existence of URIs, it just asks the agent to create and delete them like you said. The only place where we would keep track of the existence of URIs is the agent.
> 
> Isn't the previous behavior at least correct? My guess is we will just unnecessarily re-create runtime env when agent restarts, but there's no correctness issue whereas the new protocol will have the correctness issue? (correctness issue: URIs will never be deleted).

I think there's two issues: One is that runtime env creation isn't currently idempotent (conda or pip installation can fail the second time if the environment already exists). This PR fixes the idempotence issue for conda by changing the internal function to `create_conda_env_if_needed` instead of just `create_conda_env`.  I'm not sure about the status of `pip`, I'll need to do some more investigation.

The other issue is what you mentioned, URIs never being deleted.  Now that I think about it, I have a concern about the proposal to just delete all runtime_env resources when the agent restarts.  Won't this break currently running tasks and actors that depend on those resources?  For example, they might need certain `working_dir` files or `py_modules`.  Am I misunderstanding the proposal @edoakes?  

Or was the proposal based on the assumption that the "cache" is some separate directory from which resources are copied into `runtime_resources` as needed?  Currently the "cache" is just the `runtime_resources` directory itself.

rkooo567 commented:
> The other issue is what you mentioned, URIs never being deleted. Now that I think about it, I have a concern about the proposal to just delete all runtime_env resources when the agent restarts. Won't this break currently running tasks and actors that depend on those resources? For example, they might need certain working_dir files or py_modules. Am I misunderstanding the proposal @edoakes?

I think you are right here. Maybe we need additional protocols like getting used runtime env from raylet when agents restart. 

This makes me think, 
1. Is agent restart a necessary feature? Why don't we make it fate share with raylet? (so raylet is dead when agent fails). I can't think any scenario where runtime env fails & restarts other than OOM where everything can go bad.
2. Alternatively, if we'd like to keep agent restart features, why don't we just track uri counts from the raylet? And then agent just has a bunch of idempotent end points to create/delete/list resources

SongGuyang commented:
I think your discussion above is related to the issue https://github.com/ray-project/ray/issues/21695#issuecomment-1022873950.

bveeramani commented:
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

edoakes commented:
@rkooo567 fate-sharing the agent and raylet together sounds like a good idea to me, this should be very rare and it simplifies the failure handling model signficantly.

architkulkarni commented on python/ray/data/tests/test_dataset.py:
```diff
@@ -243,7 +243,7 @@ def _test_equal_split_balanced(block_sizes, num_splits):
     # Check that the expected number of rows were dropped.
     assert total_rows - expected_total_rows == total_rows % num_splits
-    # Check that all rows are unique (content check).
+    # Check that all the rows are unique (content check).
```
This trivial change was to trigger the linter since LINT was broken on master.  Unfortunately this file requires codeowner approval, @ericl or @scv119 could one of you please stamp this?

architkulkarni commented:
@rkooo567 @edoakes I agree that we should make the agent fate share with the raylet.  If we're not handling agent failures, can we merge this PR directly without putting it behind a feature flag?

@rkooo567 I addressed your comments.  The main change is to rewrite URICache with the new logic you suggested, using `set` instead of storing all the URI sizes in a `dict`.

I will make follow-up issues for the following:
- Make py_modules use a single URI like all the other plugins
- Use an abstract base class to avoid duplicated plugin code
- Change cache from random eviction to LRU
- Something to make agent fate share with Raylet (can you help give more details on what exactly needs to be done here?)

edoakes commented:
@architkulkarni I think merging this as an incremental improvements & following up on those issues SGTM. I would like to avoid us merging that something that is known to be broken though (this PR w/ agent restarts, which we currently have).

The minimal thing we can do here is clear the entire cache for each manager when the agent starts up (rm -rf resources_dir). This should be only a few lines of code, I'd suggest we do it in this PR.

architkulkarni commented:
@edoakes Sounds good, I agree we shouldn't merge something that's known to be broken.  Couple questions before I add that logic:
- What's the failure mode of the PR in its current state?  Is it just that after an agent restart, we might run into issues when calling `create()` when the resources unexpectedly already exist?  If so, I think we have the same problem before this PR, so it wouldn't be a regression.
- Won't deleting the resources directory on agent restart cause all the currently running tasks and actors that depend on those resources to fail?  Or are we saying this kind of failure is preferable to the failure above?

edoakes commented:
Good points... I think you're right that this wouldn't strictly be a regression, and adding the removal logic I suggested may indeed cause running tasks to fail. This PR should be an improvement for the majority of cases and is only problematic for the already problematic ones AFAICT, so I think we can merge it as-is. @rkooo567 thoughts?

rkooo567 commented:
If the claim is true, I think it is okay to merge as it is, but I have a question if this scenario is not a regression.

Imagine the agent is failed right after it processes DeleteURI.

Previous
- URI is properly removed eagerly. 

Current
- URI is marked as unused, but it is not removed because the disk size hasn't exceeded the limit yet
- Agent restarted, and it lost the information about uri forever 

-> in this case, isn't it going to cause leak, which is a regression? Or am I missing something here?

edoakes commented:
@rkooo567 that specific example could be considered a regression, but if you change the scenario to "agent is failed right _before_ it processes DeleteURI", we had the same issue previously.

Also, previously if the agent is failed right before it deletes a URI and then that URI is used again in the future, we would fail to create it because it already exists on disk w/o the agent knowing it.

We're basically trading of one set of undesirable behaviors on agent failure for another, but IMO this is a clear net improvement, so my recommendation is merge this and follow up w/ improving the failure handling behavior (fate sharing seems like the safe choice).

architkulkarni commented:
Here are issues for the follow-ups discussed in this PR:
- https://github.com/ray-project/ray/issues/19957
- https://github.com/ray-project/ray/issues/22059
- https://github.com/ray-project/ray/issues/22061
- https://github.com/ray-project/ray/issues/22060
